numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1931,6 +1923,419 @@ operand_different_than_broadcast: {
+ *     - (1, 0, 0) -> (1, 6, 0)
+ *     - ...
+ *   If the user limits to a range starting from 75, we use:
+ *     - (0, 7, 5) -> (0, 8, 0)  # Only 5 elements to re-align with core.
+ *     - (0, 8, 0) -> (1, 0, 0)
+ *     - ...  # continue as above
+ *
+ * This means that the data stored in the buffer has always the same structure
+ * (except when manually moved),  which allows us to fill the buffer more simply
+ * and optimally in some cases, and makes it easier to determine whether buffer
+ * content is re-usable (e.g., because it represents broadcasted operands).
+ *
+ * Best buffer and core size
+ * -------------------------
+ * To avoid having to figure out what to copy every time we fill buffers,
+ * we here want to find the outer iteration dimension such that:
+ *   - Its core size is <= the maximum buffersize if buffering is needed;
+ *   - Reductions are possible (with or without reduce mode);
+ *   - Iteration overhead is minimized.  We estimate the total overhead with
+ *     the number "outer" iterations:
+ *
+ *        N_o = full_iterator_size / min(core_size * outer_dim_size, buffersize)
+ *
+ *     This is approximately how often `iternext()` is called when the user
+ *     is using an external-loop and how often we would fill buffers.
+ *     The total overhead is then estimated as:
+ *
+ *        (1 + n_buffers) * N_o
+ *
+ *     Since the iterator size is a constant, we can estimate the overhead as:
+ *
+ *        (1 + n_buffers) / min(core_size * outer_dim_size, buffersize)
+ *
+ *     And when comparing two options multiply by the others divisor/size to
+ *     avoid the division.
+ *
+ * TODO: Probably should tweak or simplify?  The formula is clearly not
+ *       the actual cost (Buffers add a constant total cost as well).
+ *       Right now, it mostly rejects growing the core size when we are already
+ *       close to the maximum buffersize (even overhead wise not worth it).
+ *       That may be good enough, but maybe it can be spelled simpler?
+ *
+ * In theory, the reduction could also span multiple axes if other operands
+ * are buffered.  We do not try to discover this.
+ */
+static void
+npyiter_find_buffering_setup(NpyIter *iter, npy_intp buffersize)
+{
+    int nop = iter->nop;
+    int ndim = iter->ndim;
+    npy_uint32 itflags = iter->itflags;
+    NpyIter_BufferData *bufferdata = NIT_BUFFERDATA(iter);
+
+    /*
+     * We check two things here, first how many operand dimensions can be
+     * iterated using a single stride (all dimensions are consistent),
+     * and second, whether we found a reduce dimension for the operand.
+     * That is an outer dimension a reduce would have to take place on.
+     */
+    int op_single_stride_dims[NPY_MAXARGS];
+    int op_reduce_outer_dim[NPY_MAXARGS];
+
+    /*
+     * Note that this code requires/uses the fact that there is always one
+     * axisdata even for ndim == 0 (i.e. no need to worry about it).
+     */
```

Comment:
Again, only if you are making changes, maybe add:
```
 * Also, it uses that the operands have already been broadcast against
 * each other (and thus may have zero strides).
```

Reply:
Dont' feel it fits here (also moved this comment down to `best_ndim = 0`.